src/priv_appendable_data.rs

```diff
@@ -0,0 +1,521 @@
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, RustcEncodable, Debug)]
+pub struct PrivAppendedData(pub Vec<u8>);
+
+impl PrivAppendedData {
+    /// Creates a new `PrivAppendedData` encrypted with `encrypting_key`.
+    pub fn new(appended_data: &AppendedData,
+               encrypt_pub_key: &box_::PublicKey)
+               -> Result<PrivAppendedData, RoutingError> {
+        let encoded_appended_data = try!(serialise(&appended_data));
+        let encrypted_appended_data = sealedbox::seal(&encoded_appended_data, encrypt_pub_key);
+        Ok(PrivAppendedData(encrypted_appended_data))
+    }
+
+    /// Returns `AppendedData` decrypted from this item.
+    pub fn open(&self,
+                pub_key: &box_::PublicKey,
+                secret_key: &box_::SecretKey)
+                -> Result<AppendedData, RoutingError> {
+        let decipher_result = try!(sealedbox::open(&self.0, pub_key, secret_key)
+            .map_err(|()| RoutingError::AsymmetricDecryptionFailure));
+        Ok(try!(deserialise(&decipher_result)))
+    }
+}
+
+impl Decodable for PrivAppendedData {
+    fn decode<D: Decoder>(d: &mut D) -> Result<Self, D::Error> {
+        let data: Vec<u8> = try!(Decodable::decode(d));
+        if data.len() > MAX_PRIV_APPENDED_DATA_BYTES {
+            return Err(d.error("wrong private appended data size"));
+        }
+        Ok(PrivAppendedData(data))
+    }
+}
+
+/// Private appendable data.
+///
+/// These types may be stored unsigned with previous and current owner keys
+/// set to the same keys. Updates require a signature to validate.
+///
+/// Data can be appended by any key that is not excluded by the filter.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, RustcDecodable, RustcEncodable)]
+pub struct PrivAppendableData {
+    /// The name of this data chunk.
+    pub name: XorName,
+    /// The version, i.e. the number of times this has been updated by a `Post` request.
+    pub version: u64,
+    /// The keys of the current owners that have the right to modify this data.
+    pub current_owner_keys: Vec<PublicKey>,
+    /// The keys of the owners of the chunk's previous version.
+    pub previous_owner_keys: Vec<PublicKey>,
+    /// The filter defining who is allowed to append items.
+    pub filter: Filter,
+    /// The key to use for encrypting appended data items.
+    pub encrypt_key: box_::PublicKey,
+    /// A collection of previously deleted data items.
+    pub deleted_data: BTreeSet<PrivAppendedData>,
+    /// The signatures of the above fields by the previous owners, confirming the last update.
+    pub previous_owner_signatures: Vec<Signature>, // All the above fields
+    /// The collection of appended data items. These are not signed by the owners, as they change
+    /// even between `Post`s.
+    pub data: BTreeSet<PrivAppendedData>, // Unsigned
+}
+
+impl PrivAppendableData {
+    /// Creates a new `PubAppendableData` signed with `signing_key`.
+    pub fn new(name: XorName,
```

Comment:
Shouldn't this function also take `delete_data` parameter instead of initialising with a null (empty set) ? It is a signed field so i can't add it later without invalidating the signature. So if i fetch Pub/PrivAppendableData and want to change a field i will need to create a new one which is the clone of the previous one with changes in version (to re-post) and that field of choice. This will also mean cloning of deleted-data and will need to pass that to function parameter here.